src/ledger.ts

```diff
@@ -308,6 +348,35 @@ export class LedgerBitcoinInteraction extends LedgerInteraction {
     });
     return messages;
   }
+
+  /**
+   * Inheriting classes should set properties `this.isLegacySupported`
+   * and `this.isV2Supported` to indicate whether a given interaction
+   * has support for a given interaction. This method can then be called
+   * to check the version of the app being called and return whether or
+   * not the interaction is supported based on that version
+   */
+  async isSupported() {
+    if (!super.isSupported()) return false;
+    if (await this.isLegacyApp()) {
+      return this.isLegacySupported;
+    }
+    return this.isV2Supported;
+  }
+
+  /**
+   * Inheriting classes should call the super.run()
+   * as well as set the properties of support before calling their run
+   * in order to check support before calling the actual interaction run
+   */
+  async run() {
```

Comment:
Ah, the good ol' "Typescript classes can't use async constructors" obstacle. There was once a day I did it like this, but have since avoided it whenever possible. I find it better to allow for uninitialized class instances while using a `static async` method to return a class instance in a specific state. Something like `static async WithApp(): LedgerInteraction` or `static async WithSupport(version): LedgerInteraction`. Could that work here too? If so, might it make it less messy feeling than requiring a constructor call a method on the parent class? If not, perhaps this needs to be a `protected` method?

Reply:
This doesn't return an instance though. `run` methods are meant to be run on instances of the class. This particular method doesn't run anything at all. It will be superseded by inheriting classes which should call `super.run()` to do the `isSupported` check (which throws if that that interaction isn't supported). For example `LedgerExportPublicKey` will be instantiated [here](https://github.com/bucko13/unchained-wallets/blob/8a0047a8e7bba16086de61062e2c4ac2491edb72/src/index.js#L130-L155) and then `run` will be called on it. because `LedgerExportPublicKey#run` calls `super.run()` it will do the isSupported check (i was originally starting to write these manually for each interaction class and it was feeling messy.